### ISSUE TITLE
WebJobs (settings.job cron) stopped executing / no history / exception

### ISSUE THREAD

**Issue by Jabe** [projectkudu/kudu#1796]
Hi team,

some of our triggered (Cron) webjobs stopped living. One job in two Deloyment Slots stopped in September (nobody noticed), but this job and a completely different job stopped a few days ago in a more serious environment.

settings.job = `{ "is_singleton": true, "schedule": "00 */2 * * * *" }`

Symptoms:
* In https://foo.scm.azurewebsites.net/azurejobs/
 * No status
 * Last Run Time = `31.12.1 23:53:28 (less than 1ms)`
 * Go to history: No history...
* In https://foo.scm.azurewebsites.net/api/triggeredwebjobs/FooJob
 * last_run = `"status": null, "start_time": "0001-01-01T00:00:00", "end_time": "0001-01-01T00:00:00"` etc
 * Rest looks fine
* In https://foo.scm.azurewebsites.net/api/triggeredwebjobs/FooJob/history
 * Exception: `NullRef ... at Kudu.Core.Jobs.TriggeredJobsManager.CalculateETag ... GetJobHistory ... GetTriggeredJobHistory`

Then restarted SCM, no change.
Then I triggered one job by hand (`POST https://foo.scm.azurewebsites.net/api/triggeredwebjobs/FooJob/run`) and it started. Yeah!

"Touched" all jobs that way and they seem to be running just fine. One time on the dashboard they switched to "no status" again, but that might have been some kind of time-out, because all page reloads and API calls were sluggish. Might have been the jobs going crazy as well. One job execution was aborted for no apparent reason but it continued as expected 2 minutes later.

After reset:
* In https://foo.scm.azurewebsites.net/api/triggeredwebjobs/FooJob/history
 * Works again, w00t!
 * Showed normal runs until 6 days ago. Then two runs with `"status": null, "start_time": "0001-01-01T00:00:00", "end_time": "0001-01-01T00:00:00"`. Then nothing until my call to `/run`.

Any insights?

**Comment by davidebbo** [comment]
Is Always On enabled on the site? That is a requirement to use CRON.

**Comment by Jabe** [comment]
Yes, Always On is enabled.

I did some more digging. The problem appears when the `status` file is 0 byte or 193 byte of `\0`, for whatever reason this might happen. I've found instances of partially written `output_log.txt` files having trailing `\0s` and thread aborts.

Here's one of my debugging session:
```
D:\home\data\jobs\triggered\FooJob>dir
 Volume in drive D is Windows
 Volume Serial Number is A4E8-CD8D

 Directory of D:\home\data\jobs\triggered\FooJob

11/08/2015  05:59 PM    <DIR>          .
11/08/2015  05:59 PM    <DIR>          ..
07/15/2015  01:31 PM    <DIR>          201507151331151244
07/15/2015  01:32 PM    <DIR>          201507151332151505
 ... snip a lot of lines ...
07/15/2015  02:19 PM    <DIR>          201507151419151398
07/15/2015  02:20 PM    <DIR>          201507151420167396
11/08/2015  05:58 PM         1,048,606 job_prev_log.txt
11/12/2015  06:51 AM           437,348 job_scheduler.log
07/15/2015  02:20 PM             1,267 triggeredJob.lock
               3 File(s)      1,487,221 bytes
              52 Dir(s)  52,655,976,448 bytes free

D:\home\data\jobs\triggered\FooJob>dir 201507151420167396
 Volume in drive D is Windows
 Volume Serial Number is A4E8-CD8D

 Directory of D:\home\data\jobs\triggered\FooJob\201507151420167396

07/15/2015  02:20 PM    <DIR>          .
07/15/2015  02:20 PM    <DIR>          ..
07/15/2015  02:20 PM               376 output_log.txt
07/15/2015  02:20 PM                 0 status
               2 File(s)            376 bytes
               2 Dir(s)  52,655,955,968 bytes free

D:\home\data\jobs\triggered\FooJob>type 201507151420167396\output_log.txt
[07/15/2015 14:20:16 > 02facd: SYS INFO] Status changed to Initializing
[07/15/2015 14:20:16 > 02facd: SYS INFO] Run script 'run.cmd' with script host - 'WindowsScriptHost'
[07/15/2015 14:20:16 > 02facd: SYS INFO] Status changed to Running
[07/15/2015 14:20:17 > 02facd: ERR ] Thread was being aborted.
[07/15/2015 14:20:17 > 02facd: SYS INFO] WebJob process was aborted

D:\home\data\jobs\triggered\FooJob> 
```

So to repro the scheduler failure just delete the `status` file of the last run. Indeed the scheduler stops working for this job.

**Comment by davidebbo** [comment]
That's strange, When you say "the status file is 0 byte or 193 byte of \0", do you mean that these are distinct cases and that you have observed both?

In any case, we should make the logic more resilient to this so it can recover if it gets into that state for whatever reason.

**Comment by Jabe** [comment]
Yes. I've observed the following states:
* status file 0 bytes + thread abort in output_log.txt
* status file full of \0s + truncated output_log.txt with trailing \0s
* status file full of \0s + complete output_log.txt with job success

I don't know why some jobs get into this state, but it's clear that it's caused by some kind of unclear shutdown or crash. Maybe our jobs behave really naughtily sometimes and crash the webjob runner or something like that. :smile: 
I agree that the code should be a little more robust.

**Comment by BowserKingKoopa** [comment]
I have this exact problem as well. Is there a fix?

**Comment by davidebbo** [comment]
@BowserKingKoopa To help us investigate, can you share your web app name, either directly or [indirectly](https://github.com/projectkudu/kudu/wiki/Reporting-your-site-name-without-posting-it-publicly)? Thanks.

**Comment by BowserKingKoopa** [comment]
Sure.  It's http://test-drive-jobs.azurewebsites.net/ and the webjob is LeadWaker. It had stopped a while back and I didn't notice until today. I got it going again by running it once manually.

**Comment by davidebbo** [comment]
So to be clear, when you say you had the exact same problem, you mean that you found a corrupted status file as @Jabe describes above, right?

**Comment by BowserKingKoopa** [comment]
No, I have the exact problem that was described in his first post with those symptoms that he listed. I didn't look to see if my status file was 0 bytes before I ran the job manually which fixed it.

**Comment by davidebbo** [comment]
I see. We actually deployed a fix for this yesterday (63e4c4003bf856ca23d0c319b024684d7765df87). But for a site to pick it up, it needs a restart, and your site's scm site is still running the pre-fix bits. If you force a restart, that will guarantee that it won't run into this again.

**Comment by BowserKingKoopa** [comment]
Awesome. Thank you.
